### PR TITLE
Feature/gauge chart

### PR DESCRIPTION
**Pull request by ramneet-persistent**
- threshold limit added
- single threshold
- legend placement
- number of gauge input in data panel

### CODE REVIEW

File: dashboards-observability/common/constants/shared.ts [hunk 1]
```diff
@@ -158,6 +158,7 @@ export interface DefaultChartStylesProps {
   ShowLegend: string,
   LegendPosition: string,
   DefaultSortSectors: string,
+  ThresholdsMaxLimit: number
```
Comment (ruchika-narang) [inline dashboards-observability/common/constants/shared.ts]: Can we move this to Gauge chart parameters as it is not used in any other charts?

File: dashboards-observability/common/constants/shared.ts [hunk 2]
```diff
@@ -169,22 +170,8 @@ export const DefaultChartStyles: DefaultChartStylesProps = {
   ShowLegend: 'show',
   LegendPosition: 'v',
   DefaultSortSectors: 'largest_to_smallest',
+  ThresholdsMaxLimit: 1
```
Comment (ruchika-narang) [inline dashboards-observability/common/constants/shared.ts]: Move this as well to gauge chart params

File: dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_panel_option_gauge.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  );
+
+  useEffect(() => {
+    if (!vizState) {
+      setNumberOfGauges(DefaultGaugeChartParameters.DisplayDefaultGauges);
+    }
+  }, [vizState?.numberOfGauges]);
+
+  return (
+    <EuiFormRow fullWidth label="Number of gauges" helpText={`Limit number of gauges`}>
```
Comment (ruchika-narang) [inline dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_panel_option_gauge.tsx]: Can we have the helpText defined as the constant as in config panel options.

File: dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_panel_options.tsx [hunk 4]
```diff
@@ -5,6 +5,9 @@
 
 import React, { useCallback, useState, useEffect } from 'react';
 import { EuiFieldText, EuiForm, EuiFormRow, EuiTextArea, EuiAccordion } from '@elastic/eui';
+import { visChartTypes } from '../../../../../../../../common/constants/shared';
+import { ConfigPanelOptionGauge } from './config_panel_option_gauge';
```
Comment (ruchika-narang) [inline dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_panel_options.tsx]: Please check the order of imports here.

File: dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_panel_option_gauge.tsx (hunk 3 above)
Comment (ramneet-persistent) [reply dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_panel_option_gauge.tsx]: done

File: dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_panel_options.tsx (hunk 4 above)
Comment (ramneet-persistent) [reply dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_panel_options.tsx]: done

File: dashboards-observability/common/constants/shared.ts (hunk 2 above)
Comment (ramneet-persistent) [reply dashboards-observability/common/constants/shared.ts]: done

File: dashboards-observability/common/constants/shared.ts (hunk 1 above)
Comment (ramneet-persistent) [reply dashboards-observability/common/constants/shared.ts]: sure

File: dashboards-observability/public/components/visualizations/charts/financial/gauge/gauge.tsx [hunk 5]
```diff
@@ -56,17 +67,15 @@ export const Gauge = ({ visualizations, layout, config }: any) => {
...
+          ...dimensions.map((dimension: any) => data[dimension.name].slice(0, numberOfGauges)),
         ].reduce(function (prev, cur) {
           return prev.map(function (i, j) {
             return `${i}, ${cur[j]}`;
           });
         });
 
         const selectedMetricsData = [
-          ...metrics.map((metric: any) => data[metric.name].slice(0, DisplayDefaultGauges)),
+          ...metrics.map((metric: any) => data[metric.name].slice(0, numberOfGauges)),
```
Comment (spattnaik) [inline dashboards-observability/public/components/visualizations/charts/financial/gauge/gauge.tsx]: You do not need this additional destructuring here. map already returns a new array.

File: dashboards-observability/public/components/visualizations/charts/financial/gauge/gauge.tsx [hunk 6]
```diff
@@ -56,17 +67,15 @@ export const Gauge = ({ visualizations, layout, config }: any) => {
       // case 3: multiple dimensions and multiple metrics
       if (dimensionsLength && metricsLength) {
         const selectedDimensionsData = [
-          ...dimensions.map((dimension: any) =>
-            data[dimension.name].slice(0, DisplayDefaultGauges)
-          ),
+          ...dimensions.map((dimension: any) => data[dimension.name].slice(0, numberOfGauges)),
```
Comment (spattnaik) [inline dashboards-observability/public/components/visualizations/charts/financial/gauge/gauge.tsx]: Same as the comment below.

File: dashboards-observability/public/components/visualizations/charts/financial/gauge/gauge.tsx [hunk 7]
```diff
@@ -27,18 +28,28 @@ export const Gauge = ({ visualizations, layout, config }: any) => {
   // data config parametrs
   const { dataConfig = {}, layoutConfig = {} } = visualizations.data.userConfigs;
   const dataConfigTab = visualizations?.data?.rawVizData?.Gauge?.dataConfig;
-  const dimensions = dataConfigTab?.dimensions ? dataConfigTab?.dimensions : [];
-  const metrics = dataConfigTab?.metrics ? dataConfigTab?.metrics : [];
-  const dimensionsLength = dimensions.length && dimensions[0]?.name != '' ? dimensions.length : 0;
-  const metricsLength = metrics.length && metrics[0]?.name != '' ? metrics.length : 0;
+  const dimensions = dataConfigTab?.dimensions
+    ? dataConfigTab?.dimensions?.filter((i) => i.name !== '')
```
Comment (spattnaik) [inline dashboards-observability/public/components/visualizations/charts/financial/gauge/gauge.tsx]: You do not need these question marks here `dataConfigTab?.dimensions?.filter` This statement is executing, that means `dataConfigTab?.dimensions` has some values in it.

File: dashboards-observability/public/components/visualizations/charts/financial/gauge/gauge.tsx [hunk 8]
```diff
@@ -27,18 +28,28 @@ export const Gauge = ({ visualizations, layout, config }: any) => {
...
+    : [];
+  const metrics = dataConfigTab?.metrics
+    ? dataConfigTab?.metrics?.filter((i) => i.name !== '')
```
Comment (spattnaik) [inline dashboards-observability/public/components/visualizations/charts/financial/gauge/gauge.tsx]: Same as above comment.

File: dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_thresholds.tsx [hunk 9]
```diff
@@ -87,6 +88,9 @@ export const ConfigThresholds = ({
         fullWidth
         size="s"
         onClick={handleAddThreshold}
+        {...(props?.maxLimit && {
+          isDisabled: !isEmpty(vizState) && vizState.length === props?.maxLimit,
```
Comment (spattnaik) [inline dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_thresholds.tsx]: `props?.maxLimit`, this question mark may not be required, as you have checked that already in the previous condition.

**Review by spattnaik — CHANGES_REQUESTED** [review]

File: dashboards-observability/public/components/visualizations/charts/financial/gauge/gauge.tsx (hunk 5 above)
Comment (ramneet-persistent) [reply dashboards-observability/public/components/visualizations/charts/financial/gauge/gauge.tsx:78]: selectedMetricsData: array of array: it is used to get separate list of only selected metrics data from "data query response object". 

i tried to iterate on array of array as we have multiple metrics and multiple dimensions.  values are picked from inner array based on index of that inner array for both metrics and dimensions

same is case for dimensions. created separate array of concatenated labels using array of selected dimensions

@spattnaik

**Review by abasatwar — APPROVED** [review]
looks good

**Pull request merged**